src/subsidy.cpp

```diff
@@ -383,15 +383,21 @@ bool FindSubsidyIndustryCargoRoute()
 	CargoID cid;
 
 	/* Randomize cargo type */
-	if (src_ind->produced_cargo[1] != CT_INVALID && HasBit(Random(), 0)) {
-		cid = src_ind->produced_cargo[1];
-		trans = src_ind->last_month_pct_transported[1];
-		total = src_ind->last_month_production[1];
-	} else {
-		cid = src_ind->produced_cargo[0];
-		trans = src_ind->last_month_pct_transported[0];
-		total = src_ind->last_month_production[0];
+	int num_cargos = 0;
+	uint cargo_index;
+	for (cargo_index = 0; cargo_index < lengthof(src_ind->produced_cargo); cargo_index++) {
+		if (src_ind->produced_cargo[cargo_index] != CT_INVALID) num_cargos++;
```

Comment:
Why such a weird algorithm to select a random cargo type. Could as well add valid types to a std::vector and select an element from that.